src/main/java/io/github/pylonmc/pylon/base/content/magic/base/Rune.java

```diff
@@ -0,0 +1,117 @@
+package io.github.pylonmc.pylon.base.content.magic.base;
+
+import io.github.pylonmc.pylon.base.BaseKeys;
+import io.github.pylonmc.pylon.base.PylonBase;
+import io.github.pylonmc.pylon.core.config.Settings;
+import io.github.pylonmc.pylon.core.item.PylonItem;
+import io.github.pylonmc.pylon.core.item.base.PylonArrow;
+import io.github.pylonmc.pylon.core.item.base.PylonBow;
+import io.github.pylonmc.pylon.core.item.base.PylonBucket;
+import io.github.pylonmc.pylon.core.item.base.PylonTool;
+import io.github.pylonmc.pylon.core.item.base.PylonWeapon;
+import org.bukkit.FluidCollisionMode;
+import org.bukkit.block.Block;
+import org.bukkit.entity.Item;
+import org.bukkit.entity.Player;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.Listener;
+import org.bukkit.event.player.PlayerDropItemEvent;
+import org.bukkit.inventory.ItemStack;
+import org.jetbrains.annotations.NotNull;
+
+import javax.annotation.ParametersAreNonnullByDefault;
+import java.util.Collection;
+import java.util.List;
+
+/**
+ * @author balugaq
+ */
+public abstract class Rune extends PylonItem {
+    public static final double CHECK_RANGE = Settings.get(BaseKeys.RUNE).getOrThrow("check-range", Double.class);
+
+    // These can be applied with runes
+    public static final List<Class<?>> DEFAULT_APPLICABLES = List.of(
+            PylonArrow.class,
+            PylonBow.class,
+            PylonBucket.class,
+            PylonTool.class,
+            PylonWeapon.class
+    );
+
+    public Rune(@NotNull ItemStack stack) {
+        super(stack);
+    }
+
+    /**
+     * Checks if the rune is applicable to the target item.
+     *
+     * @param event  The event
+     * @param rune   The rune item, amount may be > 1
+     * @param target The item to handle, amount may be > 1
+     * @return true if applicable, false otherwise
+     */
+    @ParametersAreNonnullByDefault
+    public boolean isApplicableToTarget(PlayerDropItemEvent event, ItemStack rune, ItemStack target) {
+        PylonItem instance = PylonItem.fromStack(target);
+        if (instance == null) {
+            // Non-Pylon items are always applicable
+            return true;
+        }
+
+        if (instance instanceof RuneApplicable checker && checker.applicableToTarget(event, rune)) {
+            return true;
+        }
+
+        return DEFAULT_APPLICABLES.stream().anyMatch(clazz -> clazz.isInstance(instance));
+    }
+
+    /**
+     * Handles contacting between an item and a rune.
+     *
+     * @param event  The event
+     * @param rune   The rune item, amount may be > 1
+     * @param target The item to handle, amount may be > 1
+     */
+    @ParametersAreNonnullByDefault
+    public abstract void onContactItem(PlayerDropItemEvent event, ItemStack rune, ItemStack target);
```

Comment:
```suggestion
    public abstract void onContactItem(@NotNull PlayerDropItemEvent event, @NotNull ItemStack rune, @NotNull ItemStack target);
```
We generally avoid `ParametersAreNonnullByDefault`, we prefer explicit `NotNull`